File: Python/baekjoon/단계별로 풀어보기/스택/9012_괄호.py
# https://www.acmicpc.net/problem/9012

from sys import stdin

input = stdin.readline
for _ in range(int(input())):
    s = input().rstrip()
    while '()' in s:
        s = s.replace('()', '')
    print('NO' if s else 'YES')

File: C#/Baekjoon/스택/9012_괄호.cs
//https://www.acmicpc.net/problem/9012

using var sr = new StreamReader(new BufferedStream(Console.OpenStandardInput()));
using var sw = new StreamWriter(new BufferedStream(Console.OpenStandardOutput()));
string t;
int n = int.Parse(sr.ReadLine());
while (n-- > 0)
{
    var s = new Stack<char>();
    char d;
    t = sr.ReadLine();
    foreach (char c in t)
    {
        if (s.TryPeek(out d) && d == '(' && c == ')') s.Pop();
        else s.Push(c);
    }
    sw.WriteLine(s.Count == 0 ? "YES" : "NO");
}